Pre-review xen ffa v7 patches

jenswikl opened the pull request:
This is a dummy pull request used to get a view of the v7 patchset in progress

jenswikl committed:
xen/arm: smccc: add support for SMCCCv1.2 extended input/output registers

SMCCC v1.2 [1] AArch64 allows x0-x17 to be used as both parameter
registers and result registers for the SMC and HVC instructions.

Arm Firmware Framework for Armv8-A specification makes use of x0-x7 as
parameter and result registers.

Let us add new interface to support this extended set of input/output
registers.

This is based on 3fdc0cb59d97 ("arm64: smccc: Add support for SMCCCv1.2
extended input/output registers") by Sudeep Holla from the Linux kernel

The SMCCC version reported to the VM is bumped to 1.2 in order to support
handling FF-A messages.

[1] https://developer.arm.com/documentation/den0028/c/?lang=en

Reviewed-by: Luca Fancellu <luca.fancellu@arm.com>
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add direct request support

Adds support for sending a FF-A direct request. Checks that the SP also
supports handling a 32-bit direct request. 64-bit direct requests are
not used by the mediator itself so there is not need to check for that.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: tee: add a primitive FF-A mediator

Adds a FF-A version 1.1 [1] mediator to communicate with a Secure
Partition in secure world.

This commit brings in only the parts needed to negotiate FF-A version
number with guest and SPMC.

[1] https://developer.arm.com/documentation/den0077/e
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add defines for sharing memory

Adds defines needed for sharing using the function FFA_MEM_SHARE and
friends.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: support mapping guest RX/TX buffers

Adds support in the mediator to map and unmap the RX and TX buffers
provided by the guest using the two FF-A functions FFA_RXTX_MAP and
FFA_RXTX_UNMAP.

These buffer are later used to to transmit data that cannot be passed in
registers only.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: map SPMC rx/tx buffers

When initializing the FF-A mediator map the RX and TX buffers shared with
the SPMC.

These buffer are later used to to transmit data that cannot be passed in
registers only.

Adds a check that the SP supports the needed FF-A features
FFA_RXTX_MAP_64 / FFA_RXTX_MAP_32 and FFA_RXTX_UNMAP. In 64-bit mode we
must use FFA_RXTX_MAP_64 since registers are used to transmit the
physical addresses of the RX/TX buffers.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: send guest events to Secure Partitions

The FF-A specification defines framework messages sent as direct
requests when certain events occurs. For instance when a VM (guest) is
created or destroyed. Only SPs which have subscribed to these events
will receive them. An SP can subscribe to these messages in its
partition properties.

Adds a check that the SP supports the needed FF-A features
FFA_PARTITION_INFO_GET and FFA_RX_RELEASE.

The partition properties of each SP is retrieved with
FFA_PARTITION_INFO_GET which returns the information in our RX buffer.
Using FFA_PARTITION_INFO_GET changes the owner of the RX buffer to the
caller (us), so once we're done with the buffer it must be released
using FFA_RX_RELEASE before another call can be made.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: support guest FFA_PARTITION_INFO_GET

Adds support in the mediator to handle FFA_PARTITION_INFO_GET requests
from a guest. The requests are forwarded to the SPMC and the response is
translated according to the FF-A version in use by the guest.

Using FFA_PARTITION_INFO_GET changes the owner of the RX buffer to the
caller (the guest in this case), so once it is done with the buffer it
must be released using FFA_RX_RELEASE before another call can be made.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: move regpair_to_uint64() and uint64_to_regpair() to regs.h

Moves the two helper functions regpair_to_uint64() and
uint64_to_regpair() from xen/arch/arm/tee/optee.c to the common arm
specific regs.h. This enables reuse of these functions in the FF-A
mediator in a subsequent patch.

Reviewed-by: Michal Orzel <michal.orzel@amd.com>
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

bertrand-marquis reviewed (commented):
This is something not compliant with the spec so you will need to explain why exactly this is required.

bertrand-marquis commented on xen/arch/arm/vsmc.c:
```diff
@@ -93,7 +93,7 @@ static bool handle_arch(struct cpu_user_regs *regs)
     {
     case ARM_SMCCC_VERSION_FID:
-        set_user_reg(regs, 0, ARM_SMCCC_VERSION_1_1);
+        set_user_reg(regs, 0, ARM_SMCCC_VERSION_1_2);
```
Correct me if I am wrong but here you make Xen claim it is supporting 1.2 version but in practice all calls would still be using 1.1

jenswikl replied:
This was reviewed at https://lore.kernel.org/all/0c6d58ae-bd8d-a3d8-d8d9-2b7a5340b4dd@xen.org/

bertrand-marquis replied:
nobody commented on that part and i do not think we should claim to support 1.2 if (or until) all calls are not following the 1.2 standard otherwise a guest might have wrong expectations.

jenswikl replied:
This is needed by the Linux kernel FF-A driver.
As far as I understand we are compliant with SMCCC version 1.2. If you see anything mandatory missing please let me know and I'll add it.

bertrand-marquis replied:
1.2 is saying that the non used registers should be preserved so nothing needs to be changed unless we have calls using more registers. So it is ok.

jenswikl replied:
Thanks

bertrand-marquis commented on xen/arch/arm/include/asm/smccc.h:
```diff
@@ -265,6 +266,45 @@ void __arm_smccc_1_0_smc(register_t a0, register_t a1, register_t a2,
+    unsigned long a3;
+    unsigned long a4;
+    unsigned long a5;
+    unsigned long a6;
```
We should use a fixed size type here. I am wondering wether we should use uint64_t or register_t

jenswikl replied:
This was reviewed at https://lore.kernel.org/all/C8575668-5771-4E77-B77C-15E6E740E540@arm.com/

I'll keep it as it is for now and let you comment on the mailing list.

bertrand-marquis replied:
agree after looking at the rest it is already done this way.

bertrand-marquis commented:
so this should be mentioned on your commit message as it explains why you actually need to do this.
Also be prepared for the xen community to ask you to go and fix this in the kernel instead of trying to workaround the issue inside xen

jenswi-linaro commented:
It's not entirely clear to me how much more the commits should be split up in order to make them smaller.
These are fairly small, < 130 lines:
- xen/arm: ffa: relaxed FFA_PARTITION_INFO_GET_COUNT_FLAG support
- xen/arm: move regpair_to_uint64() and uint64_to_regpair() to regs.h
- xen/arm: ffa: support guest FFA_PARTITION_INFO_GET
- xen/arm: ffa: support mapping guest RX/TX buffers
- xen/arm: ffa: map SPMC rx/tx buffers
- xen/arm: ffa: add direct request support
- xen/arm: smccc: add support for SMCCCv1.2 extended input/output registers

This one is close to 200 lines, but I'm not sure how I should split it to make it easier to review. For instance, if I add unused static functions they are bound to cause compiler warnings.
- xen/arm: ffa: send guest events to Secure Partitions

This introduces the mediator and is quite large, but it's complete and only does the bare minimum. I could reduce the largest file, xen/arch/arm/ffa.c, quite a bit. Will that be enough? If this should be taken further how would you recommend splitting it? Tools changes and docs in one commit and the rest in another?
- xen/arm: add a primitive FF-A mediator

I'll try to take this one in steps, all structs in one commit, unfragmented sharing in one and finally fragmented sharing in the last.
- xen/arm: ffa: support sharing memory

bertrand-marquis commented:
Hi Jens,
> It's not entirely clear to me how much more the commits should be split up in order to make them smaller. These are fairly small, < 130 lines:
> 
> * xen/arm: ffa: relaxed FFA_PARTITION_INFO_GET_COUNT_FLAG support
> * xen/arm: move regpair_to_uint64() and uint64_to_regpair() to regs.h
> * xen/arm: ffa: support guest FFA_PARTITION_INFO_GET
> * xen/arm: ffa: support mapping guest RX/TX buffers
> * xen/arm: ffa: map SPMC rx/tx buffers
> * xen/arm: ffa: add direct request support
> * xen/arm: smccc: add support for SMCCCv1.2 extended input/output registers
> 
> This one is close to 200 lines, but I'm not sure how I should split it to make it easier to review. For instance, if I add unused static functions they are bound to cause compiler warnings.
> 
> * xen/arm: ffa: send guest events to Secure Partitions

All those are ok i think in term of size.

> 
> This introduces the mediator and is quite large, but it's complete and only does the bare minimum. I could reduce the largest file, xen/arch/arm/ffa.c, quite a bit. Will that be enough? If this should be taken further how would you recommend splitting it? Tools changes and docs in one commit and the rest in another?

Definitely you must split the tool changes and put it in a separate patch for 2 reasons:
- it will reduce the size
- tools and docs have an other set of reviewers

> 
> * xen/arm: add a primitive FF-A mediator
> 
> I'll try to take this one in steps, all structs in one commit, unfragmented sharing in one and finally fragmented sharing in the last.
> 
> * xen/arm: ffa: support sharing memory

Agree

jenswi-linaro commented:
Thanks

jenswikl committed:
docs: add Arm FF-A mediator

Describes a FF-A version 1.1 [1] mediator to communicate with a Secure
Partition in secure world.

[1] https://developer.arm.com/documentation/den0077/latest
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
tools: add Arm FF-A mediator

Adds a new "ffa" value to the Enumeration "tee_type" to indicate if a
guest is trusted to use FF-A.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add support for FFA_ID_GET

Adds support for the FF-A function FFA_ID_GET to return the ID of the
calling client.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: note dependency on 4k pages

Adds a BUILD_BUG_ON() to assert the dependency on 4k pages in the FF-A
mediator.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: support sharing large memory ranges

Adds support for sharing large memory ranges transmitted in fragments
using FFA_MEM_FRAG_TX.

The implementation is the bare minimum to be able to communicate with
OP-TEE running as an SPMC at S-EL1.

Adds a check that the SP supports the needed FF-A feature
FFA_MEM_FRAG_TX.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: support sharing memory

Adds support for a guest to share memory with an SP using FFA_MEM_SHARE
and FFA_MEM_RECLAIM. Only small memory regions can be shared using a
single call to FFA_MEM_SHARE are supported.

A memory region that doesn't need to be shared any longer can be
reclaimed with FFA_MEM_RECLAIM once the SP doesn't use it any longer.
This is checked by the SPMC and not in control of the mediator.

With this commit we have a FF-A version 1.1 [1] mediator able to
communicate with a Secure Partition in secure world using shared memory.
The secure world must use FF-A version 1.1, but the guest is free to use
version 1.0 or version 1.1.

Adds a check that the SP supports the needed FF-A features
FFA_MEM_SHARE_64 or FFA_MEM_SHARE_32.

[1] https://developer.arm.com/documentation/den0077/latest
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add remaining SMC function IDs

Adds the remaining SMC function IDs from FF-A 1.1 specification.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add flags for FFA_PARTITION_INFO_GET

Defines flags used for the function FFA_PARTITION_INFO_GET.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add defines for framework direct request/response messages

Adds defines for framework direct request/response messages.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add ABI structs for sharing memory

Adds the ABI structs used by function FFA_MEM_SHARE and friends for
sharing memory.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswikl committed:
xen/arm: ffa: add support to reclaim shared memory

Adds support to reclaim memory previously shared with FFA_MEM_SHARE.

Adds a check that the SP supports the needed FF-A feature
FFA_MEM_RECLAIM.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

bertrand-marquis commented on xen/arch/arm/include/asm/smccc.h:
```diff
@@ -265,6 +266,45 @@ void __arm_smccc_1_0_smc(register_t a0, register_t a1, register_t a2,
             arm_smccc_1_0_smc(__VA_ARGS__);                     \
     } while ( 0 )
+
+/**
```
Coding style, use /*

bertrand-marquis commented on xen/arch/arm/include/asm/smccc.h:
```diff
@@ -265,6 +266,45 @@ void __arm_smccc_1_0_smc(register_t a0, register_t a1, register_t a2,
+    unsigned long a17;
+};
+
+/**
```
Coding style use /*

bertrand-marquis reviewed (changes requested):
Just some minor coding style

bertrand-marquis commented on xen/arch/arm/Kconfig:
```diff
@@ -140,6 +140,17 @@ config TEE
 source "arch/arm/tee/Kconfig"
 
+config FFA
+	bool "Enable FF-A mediator support" if EXPERT
```
Please make this unsupported for now, the limitation you state on the possibility of a CPU being held for to long is a good justification for that

bertrand-marquis commented on xen/arch/arm/include/asm/ffa.h:
```diff
@@ -0,0 +1,78 @@
+/*
+ * xen/arch/arm/ffa.c
```
Wrong file name

bertrand-marquis commented on xen/arch/arm/include/asm/ffa.h:
```diff
@@ -0,0 +1,78 @@
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Please use SPDX style instead of having the full license text in the file

bertrand-marquis commented on xen/arch/arm/include/asm/ffa.h:
```diff
@@ -0,0 +1,78 @@
+
+static inline int ffa_domain_init(struct domain *d, bool ffa_enabled)
+{
+    return -1;
```
If FFA is not activated, init will always fail here.
This should return 0

bertrand-marquis commented on xen/arch/arm/vsmc.c:
```diff
@@ -32,7 +33,7 @@
 
 /* Number of functions currently supported by Standard Service Service Calls. */
-#define SSSC_SMCCC_FUNCTION_COUNT (3 + VPSCI_NR_FUNCS)
+#define SSSC_SMCCC_FUNCTION_COUNT (3 + VPSCI_NR_FUNCS + ffa_get_call_count())
```
You are calling a function in a constant, this is really not expected

bertrand-marquis commented on xen/arch/arm/vsmc.c:
```diff
@@ -196,13 +197,23 @@ static bool handle_existing_apis(struct cpu_user_regs *regs)
+{
+    uint32_t fn = fid & ARM_SMCCC_FUNC_MASK;
+
+    return fn >= 0 && fn <= 0x1fU;
```
Please create a proper define for 0x1fU

bertrand-marquis commented on xen/include/public/arch-arm.h:
```diff
@@ -318,6 +318,10 @@ struct xen_arch_domainconfig {
+#define _XEN_ARM_FLAGS_FFA        0
+#define XEN_ARM_FLAGS_FFA         (1U << _XEN_ARM_FLAGS_FFA)
+    /* IN */
+    uint32_t flags;
```
You are introducing a generic value "flags" in a public header.
I am wondering if it would not make more sense to add a:
#define XEN_DOMCTL_CONFIG_TEE_FFA 2
At the end having optee support and FFA support in the same guest does not make sense

jenswikl replied:
While I agree that it would be nice to avoid adding another field it becomes a bit confusing.
How about leaving this as is and discussing it on the mailing list when posted?

bertrand-marquis replied:
If you look in the mailing list you will see that there is a serie to add SVE which is ending in the same kind of question.
What we can do is keep it as is for now and wait for the decision on the SVE serie and then do exactly as decided for SVE :-)

bertrand-marquis replied:
as stated before, we should definitely use tee_type instead and introduce a new value for it

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; If not, see <http://www.gnu.org/licenses/>.
```
Please use SPDX instead of the whole license text

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+#define FFA_MY_VERSION          MAKE_FFA_VERSION(FFA_MY_VERSION_MAJOR, \
+                                                 FFA_MY_VERSION_MINOR)
+
+#define FFA_PAGE_SIZE                   SZ_4K
```
Why do you need a specific page size here instead of using PAGE_SIZE ?

jenswikl replied:
Because FF-A always uses 4k page size while Xen may use some other size depending on the configuration.

bertrand-marquis replied:
I understand that but you have a build_bug_on if the page_size is not 4k.
Do you expect that we could one day support a Xen page size not 4k but FFA using 4k pages ?

jenswikl replied:
Yes

bertrand-marquis replied:
ok then please add a comment here to say FFA only supports 4k page size and make the comment on the build_bug_on a bit clearer (as i commented down)

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+ * 512KB, which equals to 128 4kB pages. Due to align issue mentioned
+ * above, we need to increase this value to 129.
+ */
+#define FFA_MAX_SHM_PAGE_COUNT          129
```
Shouldn't the value here depend on page size ? something like 512K / page_size + 1 ?

jenswikl replied:
Yeah, `FFA_PAGE_SIZE`. Anyway, I'll update it.

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+
+#define FFA_HANDLE_HYP_FLAG             BIT(63, ULL)
+
+/* Memory attributes: Normal memory, Write-Back cacheable, Inner shareable */
```
We are defining an arbitrary value which i suppose come from the spec, we might want to reference where in the spec we can go to check this, same for the RW after

bertrand-marquis replied:
in fact all values here after would need to be checked against the spec, the commit message should clearly reference the spec ID and version used

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+static u16 get_vm_id(const struct domain *d)
+{
+    /* +1 since 0 is reserved for the hypervisor in FF-A */
+    return d->domain_id + 1;
```
domain_id is of type domid_t which is uint16_t so you might have an overflow

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+    return true;
+}
+
+static u16 get_vm_id(const struct domain *d)
```
any specific reason to use u16 instead of uint16_t here ?

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+    }
+}
+
+uint32_t ffa_get_call_count(void)
```
This must be converted into a proper define instead of having a function for this.
It will solve the comment made on the header

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+     * FFA_PAGE_SIZE is defined to 4k and we're currently depending on
+     * using that page size.
+     */
+    BUILD_BUG_ON(PAGE_SIZE != FFA_PAGE_SIZE);
```
If you do this you could just use PAGE_SIZE instead in the rest of the code.
BUILD_BUG_ON(PAGE_SIZE != 4k) adapting the comment

bertrand-marquis replied:
Please fix the comment here then, it is not FFA_PAGE_SIZE that is defined but FFA supporting only 4k page sizes and the code currently does support having different page sizes in xen and ffa

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -0,0 +1,372 @@
+
+#define FFA_HANDLE_INVALID              0xffffffffffffffffULL
+
+/* Framework direct request/response */
```
Please try to introduce the defines when you introduce the support for the function requiring it

bertrand-marquis reviewed (changes requested):
In the commit message you cannot reference "latest", you must give the version you used instead to make sure that someone coming in the future would look for the same document.

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -255,6 +305,66 @@ static void handle_version(struct cpu_user_regs *regs)
+    else
+        mask = GENMASK_ULL(31, 0);
+
+    src_dst = get_user_reg(regs, 1);
```
I am wondering if here we should not in fact set ourself the vm_id
A guest no knowing that it is running as a VM would probably write 0 here.
I will discuss that with other at Arm to check what they think.

bertrand-marquis replied:
in fact a guest should retrieve it using FFA_GET_ID so this is ok.

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -283,7 +399,11 @@ uint32_t ffa_get_call_count(void)
 
     if ( IS_ENABLED(CONFIG_FFA) )
-        count += 2;
+    {
```
I think here we should put the final value we will support as define instead of increasing this at each patch

jenswikl replied:
Yeah, it's painful when reworking the patches, but it is disturbing with a mismatch too when reviewing.

bertrand-marquis commented on xen/arch/arm/ffa.c:
```diff
@@ -364,6 +484,9 @@ static int __init ffa_init(void)
     printk(XENLOG_INFO "ARM FF-A Firmware version %u.%u\n",
            major_vers, minor_vers);
 
+    if ( !check_mandatory_feature(FFA_MSG_SEND_DIRECT_REQ_32) )
```
in the arm64 case, don't we expect to have support for REQ_64 ?

jenswikl replied:
We are not using FFA_MSG_SEND_DIRECT_REQ_64 in the mediator, so things work even if the SP doesn't support FFA_MSG_SEND_DIRECT_REQ_64.

bertrand-marquis reviewed (changes requested):
You should also mention in the commit message that you add support to check that the features we need from the TEE are actually supported.

jenswi-linaro commented:
Addressed review comments and rebased to be more aligned with what will be posted later.

> You should also mention in the commit message that you add support to check that the features we need from the TEE are actually supported.

I missed this comment, I'll take care of that too.

bertrand-marquis commented on xen/arch/arm/Kconfig:
```diff
@@ -141,6 +141,17 @@ config TEE
 source "arch/arm/tee/Kconfig"
 
+config FFA
+	bool "Enable FF-A mediator support" if UNSUPPORTED
```
Convention is to have (unsupported) in the name

bertrand-marquis commented on xen/arch/arm/domain.c:
```diff
@@ -715,6 +716,9 @@ int arch_domain_create(struct domain *d,
     if ( (rc = tee_domain_init(d, config->arch.tee_type)) != 0 )
         goto fail;
 
+    if ( (rc = ffa_domain_init(d, config->arch.flags)) != 0 )
```
I think the plan was to use a new value in tee_type here instead of introducing flags

bertrand-marquis commented on xen/arch/arm/domain_build.c:
```diff
@@ -4076,6 +4076,7 @@ void __init create_dom0(void)
     if ( gic_number_lines() > 992 )
         printk(XENLOG_WARNING "Maximum number of vGIC IRQs exceeded.\n");
     dom0_cfg.arch.tee_type = tee_get_type();
+    dom0_cfg.arch.flags = XEN_ARM_FLAGS_FFA;
```
same here, please use tee_type

bertrand-marquis commented on xen/arch/arm/include/asm/domain.h:
```diff
@@ -114,6 +114,10 @@ struct arch_domain
 #endif
 
+#ifdef CONFIG_FFA
+    void *ffa;
```
if we use tee_type we might also use tee here

bertrand-marquis reviewed (changes requested):
Please check to use tee_type instead of introducing flags

jenswi-linaro commented:
Updated to use tee_type instead.

jenswi-linaro commented:
Fixed the comment about "(UNSUPPORTED)". I believe I've addressed all comments now.

bertrand-marquis commented on xen/arch/arm/domain.c:
```diff
@@ -643,6 +644,7 @@ int arch_sanitise_domain_config(struct xen_domctl_createdomain *config)
     }
 
     if ( config->arch.tee_type != XEN_DOMCTL_CONFIG_TEE_NONE &&
+         config->arch.tee_type != XEN_DOMCTL_CONFIG_TEE_FFA &&
```
You should have a look into arch/arm/tee/tee.c as it would be more logical to setup ffa as a tee mediator.
This would also mean to move ffa.c into tee directory an dwould prevent most of the changes in the arm common code.
The only change that i can see is the fact that you will need to add a tee_domain_destroy but that would make sense as there is already a tee_domain_init

jenswikl replied:
`ffa_handle_call()` and `tee_handle_call()` use different SMC ID ranges, so we'd still need the changes in `handle_sssc()`. Is that OK?

bertrand-marquis replied:
Yes i think there is not other choice for that part. I currently have no idea how to make that part of the code cleaner.
Maybe during upstream review someone will have a better idea.

jenswi-linaro commented:
Updated to use the TEE mediator framework.

bertrand-marquis commented on xen/arch/arm/vsmc.c:
```diff
@@ -12,6 +12,7 @@
 #include <public/arch-arm/smccc.h>
 #include <asm/cpuerrata.h>
 #include <asm/cpufeature.h>
+#include <asm/tee/ffa.h>
```
to keep the alphabetical order this should be just before asm/tee/tee.h

bertrand-marquis reviewed (approved):
Just one minor comment with can be fixed before pushing to mailing list (for patch 1)

bertrand-marquis commented on xen/arch/arm/tee/ffa.c:
```diff
@@ -221,7 +236,11 @@ static int ffa_domain_init(struct domain *d)
+      * an overflow.
+      */
+    if ( !ffa_version || d->domain_id == UINT16_MAX)
         return -ENODEV;
```
This will prevent the guest from being created and we should say why.
I would suggest to return something different than ENODEV (EINVAL ?) and add an error print.
We might need to discuss this with the community, either domain_id 0 could be left unused by Xen or UINT16_MAX could be.

bertrand-marquis commented on xen/arch/arm/tee/ffa.c:
```diff
@@ -208,6 +258,66 @@ static void handle_version(struct cpu_user_regs *regs)
+        case FFA_MSG_SEND_DIRECT_RESP_32:
+        case FFA_MSG_SEND_DIRECT_RESP_64:
+            goto out;
+        default:
```
Here we are generating a potential infinite loop. why do we need to report back ?

bertrand-marquis reviewed (commented):
Reviewed until direct request support.
Will try to continue tomorrow but looks good :-)

jenswi-linaro commented:
I'm sorry I thought we were done for now. I've posted this as v7.

bertrand-marquis commented:
> I'm sorry I thought we were done for now. I've posted this as v7.

I think this is ok, we can switch to do the review upstream now.
